### PR TITLE
Add support for directory hashing using `IPFSHashonly`

### PR DESCRIPTION
**Pull request by angrybayblade**
## Proposed changes

This PR updates `aea.helpers.base.IPFSHashonly` tool to hash directories as well.

## Fixes

If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes and CI passes too
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: aea/helpers/ipfs/base.py [hunk 1]
```diff
@@ -92,8 +101,63 @@ def get(self, file_path: str) -> str:
         """
         file_b = _read(file_path)
         file_pb = self._pb_serialize_file(file_b)
-        ipfs_hash = self._generate_multihash(file_pb)
-        return ipfs_hash
+        return self._generate_multihash(file_pb)
+
+    def hash_directory(self, dir_path: str) -> str:
```
Comment (DavidMinarsch) [inline aea/helpers/ipfs/base.py:106]: Looks legit - can you point in docstring to the code from which this is borrowed on the IPFS golang side?

File: aea/helpers/ipfs/base.py [hunk 2]
```diff
@@ -84,6 +86,13 @@ class IPFSHashOnly:
     # according to https://pkg.go.dev/github.com/ipfs/go-ipfs-chunker#pkg-constants
 
     def get(self, file_path: str) -> str:
+        """Get the IPFS hash."""
+        if os.path.isdir(file_path):
```
Comment (Karrenbelt) [inline aea/helpers/ipfs/base.py:90]: rename to `file_path` to `path`

File: aea/helpers/ipfs/base.py [hunk 3]
```diff
@@ -92,8 +101,63 @@ def get(self, file_path: str) -> str:
...
         file_pb = self._pb_serialize_file(file_b)
-        ipfs_hash = self._generate_multihash(file_pb)
-        return ipfs_hash
+        return self._generate_multihash(file_pb)
+
+    def hash_directory(self, dir_path: str) -> str:
+        """
+        Get the IPFS hash for a directory.
+
+        :param dir_path: the dirctory path
```
Comment (Karrenbelt) [inline aea/helpers/ipfs/base.py:110]: tyop -> `directory`

**Comment by Karrenbelt** [comment]
refactor:

https://github.com/valory-xyz/open-aea/blob/294b92df8d704e169df55214c8a9d8a5f7d233ac/aea/helpers/ipfs/base.py#L126-L133
https://github.com/valory-xyz/open-aea/blob/294b92df8d704e169df55214c8a9d8a5f7d233ac/aea/helpers/ipfs/base.py#L141-L145
https://github.com/valory-xyz/open-aea/blob/294b92df8d704e169df55214c8a9d8a5f7d233ac/aea/helpers/ipfs/base.py#L194-L198

e.g.
```
def create_link(hash, size, name) -> PBLink:
    link = merkledag_pb2.PBLink()
    link.Hash = hash
    link.Tsize = size
    link.Name = name
    return link
```

File: aea/helpers/ipfs/base.py [hunk 4]
```diff
@@ -92,8 +101,63 @@ def get(self, file_path: str) -> str:
...
+
+        return str(self._hash_directory_recursively(Path(dir_path)).get("hash"))
+
+    def _hash_directory_recursively(self, root: Path) -> Dict:
+        """Hash directories recursively, starting from provided root directory."""
+
+        root_node = PBNode()
+        content_size = 0
+
+        for child in sorted(os.listdir(str(root))):
```
Comment (Karrenbelt) [inline aea/helpers/ipfs/base.py:122]: you already have a `Path` object, use `sorted(p.iterdir())`

File: aea/helpers/ipfs/base.py (hunk 1 above)
Comment (angrybayblade) [reply aea/helpers/ipfs/base.py:106]: I didn't pick anything from go-ipfs, I watched a YouTube video on IPFS file hashing and content addressing, From there I got some idea on hashing and I got some intuition from Yuri's code.

File: aea/cli/run.py [hunk 5]
```diff
@@ -215,16 +215,14 @@ def _print_hash_table(ctx: Context, aea: AEA) -> None:
                 component.public_id.author,
                 component_type.to_plural(),
                 component.public_id.name,
-                cast(str, PACKAGE_TYPE_TO_CONFIG_FILE.get(component_type.value)),
```
Comment (angrybayblade) [inline aea/cli/run.py:218]: We'll only print hashes of packages instead of yaml files

File: aea/helpers/ipfs/base.py [hunk 6]
```diff
@@ -83,17 +85,132 @@ class IPFSHashOnly:
     DEFAULT_CHUNK_SIZE = 262144
     # according to https://pkg.go.dev/github.com/ipfs/go-ipfs-chunker#pkg-constants
 
-    def get(self, file_path: str) -> str:
+    def get(self, file_path: str, wrap: bool = True) -> str:
```
Comment (angrybayblade) [inline aea/helpers/ipfs/base.py:88]: Wrapper node is optional as we discussed.

File: aea/helpers/ipfs/base.py [hunk 7]
```diff
@@ -157,20 +291,10 @@ def _generate_multihash(cls, pb_data: bytes) -> str:
         """
         multihash_bytes = cls._generate_multihash_bytes(pb_data)
         ipfs_hash = base58.b58encode(multihash_bytes)
-        return str(ipfs_hash, "utf-8")
```
Comment (angrybayblade) [inline aea/helpers/ipfs/base.py:160]: I'm not exactly sure why but using utf-8 encoding was giving out wrong hashes so I switched to str.decode() method instead. Should I investigate the reason?

**Comment by angrybayblade** [comment]
```bash
Build completed!
    _     _____     _    
   / \   | ____|   / \   
  / _ \  |  _|    / _ \  
 / ___ \ | |___  / ___ \ 
/_/   \_\|_____|/_/   \_\
                         
v1.5.0

===========================================================================================
|  ComponentId                       |  IPFSHash                                          |
===========================================================================================
|  (protocol, open_aea/signing:1.0.0)|  QmT4MmP1MYDfk9fNwRDYBnT4pp1tms62XkGW957dMrDcwb    |
|  (protocol, fetchai/default:1.0.0) |  QmRx9V1FeuUc3VGi3mmJhZk2pK6cn341zhauM9kzL9FyKJ    |
|  (connection, fetchai/stub:0.21.0) |  QmXFpXZEYWQJ4r1ZadwsBiz2fRYWp1YtbeurFPH9sLHrXF    |
|  (skill, fetchai/echo:0.19.0)      |  QmTkhfryfN2EjZy8EjwrJVfA84YEcHK1VP9Fib2bCQ1S5C    |
===========================================================================================
```

```csv
open_aea/protocols/signing,QmT4MmP1MYDfk9fNwRDYBnT4pp1tms62XkGW957dMrDcwb
fetchai/protocols/default,QmRx9V1FeuUc3VGi3mmJhZk2pK6cn341zhauM9kzL9FyKJ
fetchai/connections/stub,QmXFpXZEYWQJ4r1ZadwsBiz2fRYWp1YtbeurFPH9sLHrXF
fetchai/skills/echo,QmTkhfryfN2EjZy8EjwrJVfA84YEcHK1VP9Fib2bCQ1S5C
```

File: .github/workflows/workflow.yml [hunk 8]
```diff
@@ -365,6 +365,15 @@ jobs:
...
           make protolint_install
+
+          # install ipfs
+          wget https://dist.ipfs.io/go-ipfs/v0.6.0/go-ipfs_v0.6.0_linux-amd64.tar.gz
+          tar -xvzf go-ipfs_v0.6.0_linux-amd64.tar.gz
+          cd go-ipfs
+          sudo ./install.sh
+          cd ..
+          rm -rf go-ipfs
+          ipfs init
```
Comment (angrybayblade) [inline .github/workflows/workflow.yml:376]: Needed for the `IPFSHashOnly` tests

File: .github/workflows/workflow.yml [hunk 9]
```diff
@@ -380,6 +389,16 @@ jobs:
...
           brew install protolint
+
+          # install ipfs
+          curl -O https://dist.ipfs.io/go-ipfs/v0.6.0/go-ipfs_v0.6.0_darwin-amd64.tar.gz
+          tar -xvzf go-ipfs_v0.6.0_darwin-amd64.tar.gz
+          cd go-ipfs
+          sudo ./install.sh
+          cd ..
+          rm -rf go-ipfs
+          ipfs init
```
Comment (angrybayblade) [inline .github/workflows/workflow.yml:400]: Same as above

File: tests/test_helpers/test_ipfs/test_base.py [hunk 10]
```diff
@@ -53,3 +47,83 @@ def test_hash_for_big_file():
     data = b"1" * int(IPFSHashOnly.DEFAULT_CHUNK_SIZE * 1.5)
     my_hash = IPFSHashOnly._generate_hash(data)
     assert my_hash == VALID_HASH
+
+
+@pytest.mark.usefixtures("use_ipfs_daemon")
+@pytest.mark.skipif(
+    system() == "Windows", reason="Need to find a way to setup IPFS on windows workflow"
```
Comment (angrybayblade) [inline tests/test_helpers/test_ipfs/test_base.py:54]: Need to find a way to install IPFS using powershell

File: aea/helpers/ipfs/base.py (hunk 7 above)
Comment (Karrenbelt) [reply aea/helpers/ipfs/base.py:160]: that makes no sense to me, as the default encoding by `bytes.decode` is utf-8 (with errors='strict')

**Comment by Karrenbelt** [comment]
LGTM

File: aea/helpers/ipfs/base.py [hunk 11]
```diff
@@ -83,17 +85,132 @@ class IPFSHashOnly:
...
-        ipfs_hash = self._generate_multihash(file_pb)
-        return ipfs_hash
+        file_pb, file_length = self._pb_serialize_file(file_b)
+
+        if wrap:
+            return self.wrap_in_a_node(
+                self._generate_multihash_bytes(file_pb),
+                file_length,
+                Path(file_path).name,
+            )
```
Comment (Karrenbelt) [inline aea/helpers/ipfs/base.py:111]:
```
  if wrap:
            link_hash = self._generate_multihash_bytes(file_pb)
            link = create_link(link_hash, file_length, Path(file_path).name)
            return self.wrap_in_a_node(link)
```

same here 
https://github.com/valory-xyz/open-aea/blob/0205bc165a76fc660c049e48de9a3002841f2bc8/aea/helpers/ipfs/base.py#L127-L133

**Comment by Karrenbelt** [comment]
another stylistic thing: I would pull out the class methods (and static) methods that have nothing to do with the instance / state. Classes are overused because people default to them, whereas regular functions are perfectly fine to use.

File: aea/helpers/ipfs/base.py (hunk 7 above)
Comment (DavidMinarsch) [reply aea/helpers/ipfs/base.py:160]: Please investigate

Comment (angrybayblade) [reply aea/helpers/ipfs/base.py:160]: So michiel was right, using errors='strict' did solve the problem for some of the files, but it turned out there was also a problem with encoding which I accidentally solved while writing code to support wrapper nodes.

**Pull request merged**